### ISSUE TITLE
REQUEST [error] Error (error) while connecting to http://127.0.0.1:23119/zotxt/search.

### ISSUE THREAD

**Issue by NinKenDo64** [egh/zotxt-emacs#17]
I installed Zotero standalone, installed the zotxt add-on, and am now attempting to test out zotxt-emacs.
I installed via package-install, GNU Emacs 25.3.1 (x86_64-w64-mingw32), Windows 10 x64.

Other queries such as trying to insert the currently selected items return similar results with slightly different URLs.

Done a bit of "Googl"ing, but haven't found anybody with this issue, I might just be missing a step or something, I dunno, but I'm really at a loss.

Making a query directly in my browser, e,g., http://127.0.0.1:23119/zotxt/search?q=pragmatics%20of%20chat returns details for the item in question.

**Comment by NinKenDo64** [comment]
Problem is present whether running built-in org or ELPA org.

**Comment by egh** [comment]
Hm, that is strange. I'm guessing perhaps it is an issue with emacs/windows not being able to find 127.0.0.1 (localhost). You might try running:

```
(defconst zotxt-url-base
  "http://localhost:23119/zotxt")

```

I changed `localhost` to `127.0.0.1` at some point in the past for reasons I can't recall. We can probably change it back to `localhost` and see if that breaks anything.

**Comment by EmmanuelCharpentier** [comment]
I have the same problem. It probably resides on the emacs side. I checked than from a browser the query `http://127.0.0.1:23119/zotxt/search?q=Akaike&format=easykey` returned something sensible : 
```
[
  "akaike:1974new",
  "akaike:1974new"
]
```
(Yes, I have a double reference to this paper. Not an errror...).

I'll try de-installing/deleting/re-installing, just in case...

**Comment by egh** [comment]
@EmmanuelCharpentier Can you try `(defconst zotxt-url-base "http://localhost:23119/zotxt")`
?

**Comment by EmmanuelCharpentier** [comment]
> Can you try (defconst zotxt-url-base "http://localhost:23119/zotxt") ?

No luck : same result as before, whether I eval this before or after invocation of `org-zotxt-mode`.

**Comment by egh** [comment]
Interesting, thank you. I will install emacs on a windows VM and see if I can track this down.

**Comment by EmmanuelCharpentier** [comment]
> Interesting, thank you. I will install emacs on a windows VM andr linux see if I can track this down.

I'm running emacs and Zotero under linux (Debian testing) Our favorite scapegoat(i.e. 'Doze) isn't even party to the problem...

**Comment by NinKenDo64** [comment]
I'm unable to try that at the moment, however I forgot to mention that I actually changed the source directly to point to localhost (as opposed to 127.0.0.1) at one point. Made no difference at the time.

**Comment by caw** [comment]
I have a similar issue

Windows 10
Emacs 25.3.1
zoxtx from MELPA 20180518.147
installed the extension into standalone zotero

request from emacs gives:

REQUEST [error] Error (error) while connecting to http://127.0.0.1:23119/zotxt/search.

but typing

http://127.0.0.1:23119/zotxt/search?q=parkin

give a lovely list with the correct references...

What could I do here to help with this?

thanks so much

**Comment by EmmanuelCharpentier** [comment]
New data point : after upgrading to `emacs 26.1` (actually `27.0.50` : that's what I get from the unofficial Debian repository for emacs-snapshot), `zotxt-emacs` works as advertised.

According to the previous messages, all people having this problem ran `emacs 25.3`. this problem *might* be specific to this `emacs` version.

HTH,

**Comment by egh** [comment]
@EmmanuelCharpentier Thank you! Glad its working for you now.

**Comment by KrimpVarkie** [comment]
I have this issue on Windows 10 64 bit: first on `emacs 25.3`, then still on `emacs-26.1-x86_64`. I use `spacemacs 0.200.13`. I have tried changing `zotxt-url-base` and `zotxt-default-bibliography-style` as previously indicated, to no effect. No connection to Zotero is made, since the Zotero debug logs are empty after the attempt. However, `emacs` can still connect to http://127.0.0.1:23119/zotxt/search through the browser `eww`: hence, the issue is not that `emacs` cannot connect to `zotxt`.

Please let me know if there is anything I can do to help debug this: I would love to get this set up!

**Comment by DesiQuintans** [comment]
Hi, I also have this problem on Windows 10 64-bit, Zotero Standalone (newest version 5.0.58), and Emacs 26.1 (not Spacemacs or anything, just regular Emacs).

    > version
    GNU Emacs 26.1 (build 1, x86_64-w64-mingw32) of 2018-05-31

I've changed `zotxt-url-base` and `zotxt-default-bibliography-style` with no effect. I've also tried both versions of zotxt from MELPA (20180518.147) and MELPA Stable (0.1.34), and I've checked the Zotero Standalone client for updates (there are none). Like the previous user reported, I can also connect to the query URL with my web browser so it does exist. Happy to help debug also.

**Comment by DesiQuintans** [comment]
Okay that's strange. I have it working on my work machine (Windows 10 Enterprise 64-bit, maintained by my uni) but it's not working on my laptop (Windows 10 Home 64-bit).

- I have admin permissions on both and have been running Emacs as admin, but I've also tried running them both as a normal user.
- I made sure that the software versions on both machines matched: 
   - Emacs 26.1 of 2018-05-31
   - Zotero Standalone 5.0.59-beta.6+ffbe44a15 (I noticed that my uni machine was running the beta so I updated my laptop)
   - Firefox 64.0 64-bit
   - Zotero-to-Firefox Connector updated
   - Zotxt 5.0.2
   - Better BibTeX updated via the add-ons menu
   - All Zotero settings identical across machines

Here's a Zotero log dump from my uni computer, where I searched "dune": <https://pastebin.com/9JYVnjug> Like NinKenDo64 found, there is no log output when zotxt fails. Manually accessing the query URL <http://localhost:23119/zotxt/search?q=dune&method=titleCreatorYear&format=quickBib> in Firefox works just fine.

**Comment by tengyuhuo** [comment]
@DesiQuintans 
I have the similar situation as yours. I have it working on my surface pro 3 (Windows 10 Pro 64-bit), but it's not working on my desktop computer (Windows 10 Pro 64-bit).
The software versions and configuration are identical. Hosts files are identical on two computers.

**Comment by colliand** [comment]
I have the same issue. API responses are appearing in the browser. Emacs is reporting `REQUEST [error] Error (error) while connecting to http://127.0.0.1:23119/zotxt/items' after I attempt `C-c " i`. My references do appear inside the Helm window in emacs. After selecting one, I receive the REQUEST error above.
![](https://wwejubwfy.s3.amazonaws.com/_Minibuf-1_2019-02-02_20-28-11.png)
![](https://wwejubwfy.s3.amazonaws.com/toolchain.org_2019-02-02_20-29-02.png)

I'm using Firefox 65.0 (64 bit), Zotero Connector 5.0.53beta1, Emacs 26.1 (9.0).

**Comment by chriad** [comment]
I have the same problem. 
From the zotero debug output:

```
version => 5.0.64, 
platform => Linux x86_64, 
oscpu => Linux x86_64, 
locale => en-US, 
appName => Zotero, 
appVersion => 5.0.64, 
extensions => Zotero LibreOffice Integration (5.0.14.SA.5.0.64, extension), zotxt (5.0.2, extension)
```

emacs version:

#### System Info :computer:
- OS: gnu/linux
- Emacs: 25.2.2
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 4b195ddfc)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(haskell graphviz command-log emoji rust pandoc common-lisp-sly elisp-docstring-mode debug csv javascript emms clojure
         (mu4e :variables mu4e-installation-path "/usr/share/emacs/site-lisp")
         ruby semantic python rust html
         (markdown :variables markdown-live-preview-engine 'vmd)
         debug ipython-notebook
         (latex :variables latex-enable-folding t)
         yaml ess chrome emacs-lisp helm auto-completion ranger better-defaults git
         (org :variables org-enable-org-journal-support t org-journal-dir "~/Documents/wiki/journal/" org-journal-file-format "%Y-%m-%d" org-journal-date-prefix "#+TITLE: " org-journal-date-format "%A, %B %d %Y")
         journal pdf
         (shell :variables shell-default-height 30 shell-default-term-shell "/bin/bash" shell-default-position 'bottom)
         shell-scripts syntax-checking evil-commentary)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GSETTINGS NOTIFY ACL LIBSELINUX GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11

Firefox version:
```
firefox -V
Mozilla Firefox 65.0.1
```

**Comment by EmmanuelCharpentier** [comment]
TL;DR : The problem is definitely on emacs' side, probably in its handling of `http` communications.

I have been unable to reproduce my (extremely transient) success  with emacs-snapshot. I now think I may have been confused.

Current setup : Linux Debian testing (64 bits), emacs as packaged by Debian ("This is GNU Emacs 26.1 (build 2, x86_64-pc-linux-gnu, GTK+ Version 3.24.4) of 2019-02-03, modified by Debian"), org-mode following MELPA.

This machine (at work) is connected to a network that mandates the use of a proxy ; I set  the `http_proxy` and `https_proxy` in `/etc/profile.d/proxyconf.sh`, which intalls it in the default profile parameters.

In a temporary buffer, in org-mode with org-zotxt-mode, attempting `C-c " i`, accepting the default mode (author, year, title IIRC) and searching for `Ballantine` fails. In the `*Messages*` buffer, I get:
```
REQUEST [error] Error from parser zotxt--json-read: (json-readtable-error 60)
REQUEST [error] Error (parse-error) while connecting to http://127.0.0.1:23119/zotxt/search.
```
Searching for `http://127.0.0.1:23119/zotxt/search?q=Ballantine` in a browser (Chromium in my case, returns (correctly)  a single full entry:
```
[
  {"id":"ballantineProgressProblemsNeurological1960","type":"article-journal","title":"Progress and problems in the neurological applications of focused ultrasound","container-title":"Journal of Neurosurgery","page":"858-876","volume":"17","source":"PubMed","DOI":"10.3171/jns.1960.17.5.0858","ISSN":"0022-3085","language":"eng","author":[{"family":"Ballantine","given":"H. T."},{"family":"Bell","given":"E."},{"family":"Manlapaz","given":"J."}],"issued":{"date-parts":[[1960,9]]},"PMID":"13686380","container-title-short":"J. Neurosurg."}
]
````
as expected. The very same query i emacs' `eww` fails in an interesting way:

- Eww displays the default page of the local Apache server installed on my local machine (the displayed address is indeed the one of my 
Ethernet interface).

- The *Messages* buffer says : 

```
Using a proxy for http...
Contacting host: 127.0.0.1:23119
uncompressing publicsuffix.txt.gz...done
```
I note that *in Chromium,* attempting to request `127.0.0.1` also gives me the default Apache page. Requesting `127.0.0.1:23119` or `127.0.0.1:23119/zotxt`  get me a `no endpoint found` answer, whereas `127.0.0.1:23119/zotxt/search` gets me "q param required".

In `emacs`'s `eww`, all these request give me the default Apache page.

I am tempted to deduce that `emacs` gets confused by its proxy settings, which might be (mis-)handled.

I'll try to confirm this on a machine not using a proxy.

HTH,

**Comment by EmmanuelCharpentier** [comment]
TL;DR : it was a proxy problem in my case, indeed. Therefore, this, or a firewall problem, could be the cause for Windows users (the native Windows firewall is a bit on the paranoid side...).

On a smaller machine, with about the same software configuration, directly connected to the Internet (i. e. no proxy involved), `zotxt`, as implemented in the `org-zotxt` minor mode works as intended.

In fact, it inserts a descriptive link whose link is a `zotero:` link, the description being some restitution of the reference (style not obvious). Not so easy to use *per se*, but can be used to create a lot of useful things (e. g. a dynamically-created BibTeX file, or an exporter to `docx|odt` using native zotero plugin fields, etc...).

OTOH, I haven't been able to use `zot-easykey-*` functions. Probably because I am not familiar with them : a `zotxt` guide is sorely needed...

Question : does anybody know how to tell `emacs` **NOT** to proxy requests to `localhost:23119` ?

HTH,

**Comment by egh** [comment]
@EmmanuelCharpentier Thank you so much for this!! I will take a look on Monday.

**Comment by EmmanuelCharpentier** [comment]
Dear Erik,

1) Answering my own question (in order to help future Googlers with the same issue) : 

> does anybody know how to tell emacs NOT to proxy requests to localhost:23119 ?

One can either
- define the `NO_PROXY` (or `no_proxy`) environment variable to a comma-separated list of hosts, domains and URLs not to be proxified, or
- use the `url-proxy-services` of the `url` emacs library to the same end (probably in the `.emacs` starttup file).

This is by no way obvious to find, since it is not in the `emacs` manual but is [buried](https://www.gnu.org/software/emacs/manual/html_node/url/Proxies.html#Proxies) in the library's own manual...

Since I share my `.emacs` file between several machines, not all of them behind a proxy firewall, I used the first solution (where this can be set up in `.profile`), with success.

2) The way to use this facility isn't obvious, and I may have a couple of suggestions. I will open a separate issue for this.

3) I'd be rather interested to know if the solution I propose works for the Windows users who reported the problem. If so, this issue could be closed.

4) What became of the look you planned to do ?

**Comment by zsaur** [comment]
Had the same problem here and found the solution. It had nothing to do with this repo. I was following this guide http://www.mkbehr.com/posts/a-research-workflow-with-zotero-and-org-mode/ and there was some emacs configuration code that I just copied and pasted without paying much attention:
```emacs-lisp
;; Change citation format to be less cumbersome in files.
;; You'll need to install mkbehr-short into your style manager first.
(eval-after-load "zotxt"
'(setq zotxt-default-bibliography-style "mkbehr-short"))
```

It turns out this part was causing the problem and removing it solved the problem.

I am suspecting that some of the people here followed the same guide and did the same mistake

**Comment by egh** [comment]
@zsaur That's very odd - I'm not sure why that would cause on issue, but it's good to know. Thank you!

**Comment by proudot** [comment]
Hi folks, I am reporting the same issue on Win10 with emacs 26.1. I am behind a firewall by my url-proxy-services are set up correctly for use with MELPA.  Using 

```
(defconst zotxt-url-base
  "http://localhost:23119/zotxt")
```

didn't help. I didn't copy MkBehr tutorial. Any ideas ?

**Comment by egh** [comment]
Hi all,
Another user reported a bug https://github.com/egh/zotxt/issues/12 that mentioned the `--compressed` option not being supported in the `curl` on windows. I wonder if someone might try:
```
(setq request-backend 'url-retrieve)
```
and see if that corrects the issue for you?

**Comment by plandes** [comment]
I am now having this issue as well.  I've tried all the suggestions above.

Is there any update?

Here's my version configuration:
- Zotxt 5.0.3
- Zotero 5.0.80
- Orgmode 9.1.9
- Emacs 26.2

**Comment by TshuW** [comment]
I originally opened the link with `org-open-at-point`, but later I realized that I should use`org-zotxt-open-attachment` and point out here in case someone makes the same mistake as me.

**Comment by egh** [comment]
Hi @plandes , sorry for the delay, I'm hoping to get my hands on a mac in the next day or two to look into this!

**Comment by pbienst** [comment]
Windows 10 setup here. For me the fix (setq request-backend 'url-retrieve) is needed when trying to import data from zotero. 

However, when clicking the link to open the item in zotero, I get a request-default-error-callback http://127.0.0.1:23119/zotext/select error

**Comment by egh** [comment]
@plandes I just tested on:

macos 10.14.6
emacs 26.3
zotero 5.0.80
latest org mode from melpa
zotxt 5.0.3

and I was unable to reproduce any issues. Is it possible to set:

`(setq request-log-level 'debug)`

and paste the content in the ` *request-log*` buffer here? (Note the space in front of the buffer.) Thank you!

**Comment by pbienst** [comment]
```
[warn] request--url-retrieve-callback: (error http 500)
[debug] request--callback: UNPARSED
HTTP/1.0 500 Internal Server Error
X-Zotero-Version: 5.0.80
X-Zotero-Connector-API-Version: 2
Content-Length: 0


[error] request--callback: peculiar error: 500
[debug] request--callback: executing error
[error] request-default-error-callback: http://127.0.0.1:23119/zotxt/select error
[warn] request--url-retrieve-callback: (error http 500)
[debug] request--callback: UNPARSED
HTTP/1.0 500 Internal Server Error
X-Zotero-Version: 5.0.80
X-Zotero-Connector-API-Version: 2
Content-Length: 0


[error] request--callback: peculiar error: 500
[debug] request--callback: executing error
[error] request-default-error-callback: http://127.0.0.1:23119/zotxt/select error
```

**Comment by egh** [comment]
Thank you! I'll look into this further tonight.

**Comment by plandes** [comment]
@egh I set the debug level as you requested, and I still get this:
```log
REQUEST [error] Error (error) while connecting to http://127.0.0.1:23119/zotxt/search.
````

I also tried to connect directly to this bound port with `netcat` and get this:
```shell
$ nc -v 127.0.0.1 23119
found 0 associations
found 1 connections:
     1:	flags=82<CONNECTED,PREFERRED>
	outif lo0
	src 127.0.0.1 port 53851
	dst 127.0.0.1 port 23119
	rank info not available
	TCP aux info available

Connection to 127.0.0.1 port 23119 [tcp/*] succeeded!
```

Hope this is helpful in debugging as its all I have to give.  Let me know what else I can do.

Thanks very much in advance.

**Comment by egh** [comment]
@pbienst For your issue, there was an upstream change to Zotero that affected the `select` function. I have created a new release to fix this:

https://github.com/egh/zotxt/releases/tag/v5.0.4

**Comment by egh** [comment]
@plandes After setting the debug level, can you report what is in the ` *request-log*` buffer? (There is a space before the first `*` in that buffer name).

**Comment by pbienst** [comment]
Works now, thanks!

**Comment by egh** [comment]
@pbienst Great!

**Comment by plandes** [comment]
@egh  Ah, I see the buffer now.  Here are its contents:
```
[debug] REQUEST
[debug] Run: curl --silent --include --location --compressed --cookie ~/.emacs/request/curl-cookie-jar --cookie-jar ~/.emacs/request/curl-cookie-jar --write-out \n(:num-redirects %{num_redirects} :url-effective "%{url_effective}") http://127.0.0.1:23119/zotxt/search?q=Towards&method=titleCreatorYear&format=quickBib
[debug] REQUEST--CURL-CALLBACK event = finished

[debug] REQUEST--CURL-CALLBACK proc = #<process request curl>
[debug] REQUEST--CURL-CALLBACK buffer = #<buffer  *request curl*>
[debug] REQUEST--CURL-CALLBACK symbol-status = nil
[debug] REQUEST--CALLBACK
[debug] (buffer-string) =
HTTP/1.0 404 Not Found
X-Zotero-Version: 5.0.80
X-Zotero-Connector-API-Version: 2
Content-Type: text/plain

No endpoint found

[debug] REQUEST-RESPONSE--CANCEL-TIMER
[debug] -CLEAN-HEADER
[debug] -CUT-HEADER
[debug] error-thrown = (error http 404)
[debug] -PARSE-DATA
[debug] parser = zotxt--json-read
[debug] data = nil
[debug] symbol-status = error
[debug] Executing error callback.
[error] Error (error) while connecting to http://127.0.0.1:23119/zotxt/search.
```

**Comment by egh** [comment]
@plandes If the endpoint is not found, that means that zotxt failed to load correctly. Can you restart Zotero and check your zotxt add on installation? Thanks!

**Comment by plandes** [comment]
@egh I found the extension wasn't installed (even after installing around the time of the first problem).  So then I went to install it again and restarted and found it still missing.  I then went to install again, and judiciously searched and confirmed the extension existed before restarting.  The third time was a charm.  It's either a blip in some quantum field, parallel universe phenomena and <insert any story from Hitchhiker's Guide to the Galaxy>, but for now it seems to be working.

Thanks for your help and sorry for the static.

**Comment by egh** [comment]
@plandes Strange. Perhaps there is a problem with the installation. I'll take a look, but I'm glad it's working for you now!

**Comment by plandes** [comment]
@egh is there a Zotero error log file somewhere on the filesystem we can look at?

**Comment by egh** [comment]
@plandes you should be able to look at the debug log by following the instructions here: https://www.zotero.org/support/debug_output

**Comment by plandes** [comment]
I don't see anything so I suppose it clears on restart and it is too late.  I'll do this next time.  Thanks again for your help Erik!

**Comment by egh** [comment]
I have improved the testing on zotxt-emacs and zotxt. I have added a test endpoint to ensure that basic zotxt functionality is available. I advise anyone with this issue to upgrade zotxt and zotxt-emacs and see if the issue can be reproduced. I'm going to close this issue, but please do open a new issue if you see one! Thanks everybody for the bug reports.

**Comment by AtomicNess123** [comment]
I have this problem. Any solution?
